File: 3394_check_if_grid_can_be_cut_into_sections.cpp
class Solution {
public:
    bool checkValidCuts(int n, vector<vector<int>>& r) {
        auto z = [&](vector<pair<int,int>> v) {
            sort(v.begin(), v.end());
            int c = 1, m = v[0].second;
            for (auto [a, b] : v) if (a >= m) m = b, ++c; else m = max(m, b);
            return c >= 3;
        };
        vector<pair<int,int>> x, y;
        for (auto& q : r) x.emplace_back(q[0], q[2]), y.emplace_back(q[1], q[3]);
        return z(x) || z(y);
    }
};

File: 3394_check_if_grid_can_be_cut_into_sections.java
class Solution {
    public boolean checkValidCuts(int n, int[][] r) {
        return x(r) || y(r);
    }
    boolean x(int[][] r) {
        List<int[]> z = new ArrayList<>();
        for (int[] q : r) z.add(new int[]{q[0], q[2]});
        z.sort(Comparator.comparingInt(a -> a[0]));
        int m = z.get(0)[1], s = 1;
        for (int i = 1; i < z.size(); i++) {
            if (z.get(i)[0] >= m) { s++; m = z.get(i)[1]; }
            else m = Math.max(m, z.get(i)[1]);
        }
        return s >= 3;
    }
    boolean y(int[][] r) {
        List<int[]> z = new ArrayList<>();
        for (int[] q : r) z.add(new int[]{q[1], q[3]});
        z.sort(Comparator.comparingInt(a -> a[0]));
        int m = z.get(0)[1], s = 1;
        for (int i = 1; i < z.size(); i++) {
            if (z.get(i)[0] >= m) { s++; m = z.get(i)[1]; }
            else m = Math.max(m, z.get(i)[1]);
        }
        return s >= 3;
    }
}
